cli/cmd/agent_list.go

```diff
@@ -0,0 +1,208 @@
+// Copyright:: Copyright 2022, Lacework Inc.
+// License:: Apache License, Version 2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package cmd
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"github.com/lacework/go-sdk/api"
+)
+
+var (
+	agentListCmdState = struct {
+		// The available filters for the agent list command
+		AvailableFilters CmdFilters
+
+		// List of filters to apply to the agent list command
+		Filters []string
+	}{}
+
+	agentListCmd = &cobra.Command{
+		Use:     "list",
+		Aliases: []string{"ls"},
+		Short:   "List all hosts with a running agent",
+		Long: `List all hosts that have a running agent in your environment.
+
+You can use 'key:value' pairs to filter the list of hosts with the --filter flag.
+
+    lacework agent list --filter 'os:Amazon Linux' --filter 'tags.VpcId:vpc-72225916'
+
+**NOTE:** The value can be a regular expression such as 'hostname:db-server.*'
+
+The available keys for this command are:
+` + stringSliceToMarkdownList(agentListCmdState.AvailableFilters.GetFiltersFrom(api.MachineDetailEntity{})),
+		RunE: listAgents,
+	}
+)
+
+func init() {
+	agentListCmd.Flags().StringSliceVar(&agentListCmdState.Filters, "filter", []string{},
+		"filter results by key:value pairs (e.g. 'tags.Name:worker-.*')",
+	)
+}
+
+func listAgents(_ *cobra.Command, _ []string) error {
+	var (
+		progressMsg = "Fetching list of agents"
+		response    = &api.MachineDetailsEntityResponse{}
+		now         = time.Now().UTC()
+		before      = now.AddDate(0, 0, -7) // 7 days from ago
+		filters     = api.SearchFilter{
+			TimeFilter: &api.TimeFilter{
+				StartTime: &before,
+				EndTime:   &now,
+			},
+		}
+	)
+
+	if len(agentListCmdState.Filters) != 0 {
+		progressMsg = fmt.Sprintf(
+			"%s with filters (%s)",
+			progressMsg, strings.Join(agentListCmdState.Filters, ", "),
+		)
+		filters.Filters = []api.Filter{}
+		for _, tag := range agentListCmdState.Filters {
+
+			kv := strings.Split(tag, ":")
+			if len(kv) != 2 {
+				cli.Log.Warnw("malformed tag filter, ignoring",
+					"tag", tag, "expected_format", "key:value",
+				)
+				continue
+			}
+
+			cli.Log.Infow("adding filter", "key", kv[0], "value", kv[1])
+			filters.Filters = append(filters.Filters, api.Filter{
+				Field:      kv[0],
+				Expression: cli.lqlOperator, // @afiune we use rlike to allow user to pass regex
+				Value:      kv[1],
```

Comment:
Should we check for empty string here?

ERROR unable to list agents via MachineDetails search: 
  [POST] https://customerdemo.lacework.net/api/v2/Entities/MachineDetails/search
  [400] Invalid filters: expression 'rlike' requires non-empty 'value' field